common/src/main/java/com/microsoft/identity/common/internal/dto/Credential.java

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.microsoft.identity.common.internal.dto;
+
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * This is a generic credential schema that should be used as a reference to define schemas for
+ * specific credential types.
+ */
+public class Credential {
```

Comment:
TODO: Need to figure out proper handling for additional json fields.

Reply:
@oldalton Are you maintaining the full raw-response from AAD and writing that to the cache as a special field on the cache item? How are you treating JSON payloads (either read from disk or network) that contain fields you don't recognize?

Reply:
For the first question, no, I'm not writing full raw AAD response. For additional JSON payloads read from disk, they're just stored internally and written back (JSON payload is represented as a map after reading and new values are overriding old ones). For additional fields coming from server, it's the same story, but those are not written to cache. What's the behavior on Android?

Reply:
Currently neither ADAL, nor MSAL, nor Broker has persisted any extra fields. They're ignored and discarded. @shoatman and I were discussing the morning of doing something approximating the following (using simple classes and JSON objects to express the idea):

Suppose we have a simple class:

```java
class Car {
    @SerializedName("make")
    private String mMake;

    @SerializedName("model")
    private String mModel;

    @SerializedName("model_year")
    private int mModelYear;

    private transient Map<String, JSONObject> mAdditionalFields;
}
```

and our pretend web service returns the following payload:

```json
{
  "make": "Honda",
  "model": "Civic",
  "model_year": 2017,
  "door_count": "4"
}
```

Then, under the proposed idea, we'll parse the object such that...

```java
// Parse the response to create the Car object
Car car = new Gson().fromJson(jsonResponse, Car.class);

// Query its fields...
car.mMake.equals("Honda"); // returns true
car.mModel.equals("Civic"); // returns true
car.mModelYear == 2017; // returns true
car.mAdditionalFields.get("door_count"); // returns JSONObject (containing String value)

// Take note...
car.mAdditionalFields.size() == 1; // returns true
car.mAdditionalFields.get("make"); // returns null
```

Seem alright @oldalton @shoatman ?

Reply:
Still experimenting with this idea so subject to change but the idea would be to support complex actions on additional fields if it is a nested object.

For instance, this could be handled semi-gracefully (to re-use the car example):
```json
{
	"make": "Honda",
	"model": "Civic",
	"model_year": 2017,
	"door_count": "4",
	"engine": {
		"cylinders": 4,
		"displacement": 2.0
	}
}
```

We could then query this additional field
```java
car.mAdditionalFields.get("engine").getInt("cylinders") == 4; // returns true
```

I don't know yet if I'm going to take this approach, as it may just be unnecessary